README.md

```diff
@@ -0,0 +1,396 @@
+
+Apps represent the final binaries defined by the agent.
+Each "flavor" of agent is defined in a different sub-package of `cmd/`.
+
+Apps are formulaic and should not contain any complex logic.
+Their job is to parse command-line options, set up an `fx` App, and run it.
+
+For parameterized components, the app must also supply the parameters:
+
+```go
+app = fx.New(
+    foo.Module,
+    fx.Supply(foo.ModuleParams{
+        MaxFoos: 13,
+    }),
+    ...
+)
+```
+
+### Component Dependencies
+
+Component dependencies are automatically determined from the arguments to a component constructor.
+For example, a component that depends on the log component will have a `logs.Component` in its argument list:
+
+```go
+import "github.com/djmitche/dd-agent-comp-experiments/comp/util/log"
+
+func newThing(..., log log.Component, ...) Component {
+    return &thing{
+        log: log,
+        ...
+    }
+}
+```
+
+### Component Bundles
+
+Many components naturally gather into larger areas of the agent codebase, such as DogStatsD.
+In many cases, these components are not intended for use outside of that area.
+These components should be defined in an `internal/` package, and included in a `Module` definition in the parent package.
+This single `Module` can then be included by apps that require that functionality.
+
+### Testing
+
+Testing for a component should use `fxtest` to create the component.
+This focuses testing on the API surface of the component against which other components will be built.
+Per-function unit tests are, of course, also great where appropriate!
+
+Here's an example testing a component with a mocked dependency on `other`:
+
+```go
+func TestMyComponent(t *testing.T) {
+    var comp Component
+    var other other.Component
+    app := fxtest.New(t,
+        Module,              // use the real version of this component
+        other.MockModule,    // use the mock version of other
+        fx.Populate(&comp),  // get the instance of this component
+        fx.Populate(&other), // get the (mock) instance of the other component
+    )
+
+    // start and, at completion of the test, stop the components
+    defer app.RequireStart().RequireStop()
+
+    // cast `other` to its mock interface to call mock-specific methods on it
+    other.(other.Mock).SetSomeValue(10)                      // Arrange
+    comp.DoTheThing()                                        // Act
+    require.Equal(t, 20, other.(other.Mock).GetSomeResult()) // Assert
+}
+```
+
+If the component has a mock implementation, it is a good idea to test that mock implementation as well.
+
+# Conventions
+
+## Package paths
+
+ * `cmd/<appname>/main.go` -- entrypoint for each app
+ * `pkg/<pkgnname>/...` utility types and functions, ["plain old data"](https://en.wikipedia.org/wiki/Passive_data_structure) types
+ * `pkg/util/<utilname>/...` utility packages that are not in the form of a component
+ * `comp/...` components
+ * `comp/<bundlename>/...` component bundles
```

Comment:
I need to work on this more.  The idea is that "logs-agent" is a bundle, and "DogStatsD" is a bundle, and "trace-agent" is a bundle, and maybe "core" is a bundle (comp/util/logs, comp/flare, comp/health, comp/status).

Reply:
I am not sure we need bundles either.
For instance, I don't believe `trace-agent` should be a bundle, at least in terms of naming. We want to be able to run any workload in any process, so we may have a component named `TraceReceiver`.
Maybe the `core` bundle makes sense (like a special, non-functional, bundle).

Reply:
Let's say I want to run everything in a single process, the equivalent of `trace-agent`, `process-agent` and `core agent (logs+metrics)`.

That will be known at runtime, at build time we'll build a single binary. Maybe it makes sense to have `bundles` to illustrate top-level components and select those we want to run, but the grouping looks a bit arbitrary.

Like is the container metrics collection a top-level component (e.g. bundle), or will the corecheck using it be one?
While I'm thinking about it, do you have an idea of how to represent core checks in this architecture?

Reply:
I think the grouping may change over time, and Conway's law says it will start to follow the organization -- so, container metrics will gravitate toward other components owned by the team that maintains it.

I hadn't given much though to the structure of checks and collectors and check runners -- do you want to take a swing at that?